### DISCUSSION TITLE
Introduction to Source Separation - Tutorial

### DISCUSSION THREAD

**Discussion by ethman** [nussl/nussl#195] [General]
Hi everyone!

At this year's ISMIR, @pseeth, Justin Salamon, and I presented a tutorial on getting started working with and training modern deep learning-based models. It talked about using [Scaper](https://github.com/justinsalamon/scaper) and nussl together to build a modern pipeline for training, eval, and use.

[Check out the tutorial here.](https://source-separation.github.io/tutorial/)

If you're looking to get started, the tutorial is a good way to dive into this tech and gain familiarity with the nussl codebase.

**Comment by jaslineljy** [comment]
Hi, @ethman i have viewed your https://source-separation.github.io/tutorial/landing.html. It is quite interesting and I have learnt some useful knowledge. However, as a beginner who is new to this, there are still some questions I would like to ask:

1)  Under the `Put all Together section` code, the tutorial taught about how to separate vocals. If I would want to separate to drums, bass, vocals and others, does that mean that i don't use `sum_sources = nussl_tfm.SumSources` at all? In that case, does that mean that when plotting the magnitude spectrogram i have to change 1 to 4 in the code: 
`index_sources = nussl_tfm.IndexSources('source_magnitudes', 1)` 
Are there any other changes that i am supposed to make or be aware of? Like for the evaluation of sources SDR.

2) Regarding the interaction interface with gradio, i assume that after typing `separator.model.save('checkpoints/best.model.pth')` i should be able to see the file in the folder but i did not see the saved model file on colab. Is there a way to actually download the save model file too?


3) At the last part of the `Put all Together section` code, how do we exactly import the saved model and import for interaction? I HPSS code and make use of the last line of the tutorial code`separator.interact(share=True, source='microphone')` and typed the following:
>> import nussl
nussl.separation.primitive.HPSS(nussl.AudioSignal()).interact(share=True, source='microphone')

If i have another model saved as .th format (not from tutorial code), how can i make use of the code above to create an interaction with nussl with this arbitrary model file that i have.


4) I would like to know whether under the `Put all Together section` , does the following code means that it uses the full 150 songs from MUSDB18?
>> data.prepare_musdb('~/.nussl/tutorial/')

How is this different from the code stated in `MUSDB dataset section` and `Coding up model architecture section`?
>>musdb = nussl.datasets.MUSDB18(download=True)

I attempted to download these 2 sets of files because i was curious what was the difference. However, i got a code error on downloading these datasets. So it would be great if you could enlighten me what is the difference and also provide advice on how I can actually import a custom dataset or my downloaded MUSDB dataset that i have in my hard disk to tutorial code as I am experiencing an error.  Can i also confirm whether the custom dataset also needs to match the number/ type of sources input during training? For example, if training is vocal, drums, bass, and others can my test set be piano and drums only? What changes do I have to make to ensure that the tensors are of the correct size?

5) Just to confirm, is it the activation function for masking has to always be sigmoid? Can it be relu?

Look forward to your reply. Thank you.

**Reply by ethman** [reply]
Hey thanks for migrating your questions here. It helps us stay organized.


1. All you need to do is switch the that number to the correct index that you car about. The sources are in alphabetical order, so `0: Bass, 1: Drums, 2: Other, 3: Vocals` for MUSDB18. Just switch the index to get the source you want.
2. Try saving the model in the home directory instead (`~/`)? I honestly don't know the ins and outs of how Colab lets users access its file system, but maybe this will work better. Sorry I can't be of more help.
3. So the HPSS object is incompatible with any trained deep model. For that you'll need to use a `DeepMaskEstimation` object. What you want is `seperator = nussl.separation.deep.DeepMaskEstimation(signal, model_path)`.
4. Nope, that only gives you the preview files which are 7 seconds long. At this point in time you cannot download MUSDB18 through the command line (though that [will be changing soon](https://twitter.com/faroit/status/1360885033008041988?s=20)). Here's the link to download the full dataset: https://zenodo.org/record/1117372. I don't know how to get it onto colab though. Maybe curl?
5. The activation can be whatever you like. Sigmoid is used because it's bounded between [0.0, 1.0], which is nice for making sure the mask is always well behaved, but plenty of algorithms use ReLU (Open-Unmix uses ReLU iirc).

**Reply by jaslineljy** [reply]
Hi, I am still confused on how to move on based on your previous reply
1) I understand that but changing the index from `index_sources = nussl_tfm.IndexSources('source_magnitudes', 1)` will allow me to get the 1 source that i want (in the tutorial case is vocals). However, how do i stack all the sources together so that at the end of the day i am able to **simultaneously** extract Bass, Drums Other, and vocals at once. I tried the 2 following ways:
a) Comment out `sum_sources = nussl_tfm.SumSources([['bass', 'drums', 'other']])` since i do not want to combine the sources but obtain all the 4 separate sources at the end. 
b) Sum each of the own source to itself `sum_sources = nussl_tfm.SumSources([['bass'], ['drums'], ['other'],['vocals']])`

Though visually it looks correct when i did `viz.show_sources(item['sources'])` i got error in execution as I am confused to how
i should manipulate the following code:
>a) plt.imshow(item['source_magnitudes'][..., 1][..., 0], aspect='auto', origin='lower')
>b) index_sources = nussl_tfm.IndexSources('source_magnitudes', 1)
>c) plt.imshow(item['source_magnitudes'][..., 0, 0], aspect='auto', origin='lower')

To add on, You mentioned in `Coding up model architectures section` that: 

![image](https://user-images.githubusercontent.com/44189034/109245562-b3261680-781b-11eb-975b-1dc854d3b21d.png)

So i am curious how exactly i can stack and what are the affected lines of code that i have to take care of by doing this. 


2) Regarding the use of `seperator = nussl.separation.deep.DeepMaskEstimation(signal, model_path)` which is similar to the following stated under the `Put it all together section`
![image](https://user-images.githubusercontent.com/44189034/109243535-11e99100-7818-11eb-8238-7e56d017dc45.png)
a) May i know what does `nussl.AudioSignal()` or `signal` refer to? I am assuming it refers to the signal from the unknown file source that we can input to separate using the model. Since there isn't any unknown file, it leads to the error of `'input_audio_signal has no data!'` is that right?
b) If that is the case, doesn't that mean that I should put a file link of the song instead so that it could detect the signal from the song file. And how should i go about writing the code to input the file song directory as the input signal?

3) You mention that HPSS object is incompatible with any trained deep model. I am assuming that applies to the tutorial model as well? Because I really like how there is the interaction with gradio since it makes it clearer of how everything works at a glance. I noted that you stated that another method to use other than HPSS is the following
>>import nussl
dummy_signal = nussl.AudioSignal()
ft2d = nussl.separation.primitive.FT2D(dummy_signal)
ft2d.interact()

May I ask
a) how I can place my model in the above code? Since i don't see that there is a way to place `seperator = nussl.separation.deep.DeepMaskEstimation(signal, model_path)` you mentioned previously. 
b) For `nussl.AudioSignal()`, since it uses the gradio interface, i am assuming that `nussl.AudioSignal()` will immediately take my random input audio file and placed as the dummy_signal? Do i need to specify any audio source directory? 

4) For the checkpoints and importing/exporting of data with Colab, I will try to look into it. Just curious, if i manage to import the MUSDB dataset (shld it be the compressed one or wav?) do i still need to go through Scaper?

Look forward to your prompt reply.

**Reply by ethman** [reply]
Hi,

1. Whoops! I'm sorry. That's my fault; I misinterpreted what you were asking with your initial question. If you want to make masks for all four sources, you have to remove the `IndexSources` and `SumSources` transforms from your data loader. By default the MUSDB18 data loader will return all of the sources, and you need to use `IndexSources` or `SumSources` to reduce that, so if you don't use either you will get all four sources.
If you do this you'll also need to update your network so that it produces 4 sources or masks (this all depends on how you set up your network). There are so many ways to set up the network that it's hard to give you any more specific advice. You need to train it with 4 sources. In other words, you can't take a model trained to separate just Vocals and then try to separate all four sources with it.

2. You don't have to worry about those warnings upon initialization, so you can ignore them at that step. Those are just _warnings_, not _exceptions_ (exceptions cause the program to crash, warnings do not). In general, warnings just tell you that _it is possible_ that you have something set up incorrect. And in those two lines technically we haven't finished setting up so the warnings are valid, but the rest of the setup happens a few lines later, which is why it's okay for us to ignore those warnings when we see them. However, to separate audio you will need to give the `separator` model a populated audio signal. You can see where this happens a few lines below the image you posted (highlighted in yellow below): 
<img width="689" alt="image" src="https://user-images.githubusercontent.com/2359841/109435301-be429780-7a11-11eb-8d4b-13b6ae8d98a5.png">
`item['mix']` is the mixture from the dataset item. Then in the line below the yellow highlighted one, we do the actual separation by calling `estimates = separator()` which returns a set of source estimates from our model.


3. It might be helpful to think of `HPSS` and `FT2D` as _different models_ than `DeepMaskEstimation`. So if you want to use your trained model (e.g. `my_model = nussl.separation.deep.DeepMaskEstimation('...')`) interactively, it doesn't make sense to do `ft2d.interact()` and expect it to be the results of `my_model`. To make `my_model` interactive, you need to call `interact()` on the `my_model` object like this `my_model.interact()`. This will work for any separation model in nussl: Repet, RepetSim, DeepClusteringEstimation, DeepMaskEstimation, Duet, and so on. 

4. As with our other thread on MUSDB, you will still have to do the preprocessing to get it a set of `.wav` files. You can use Scaper if you want, but that is up to you.

**Reply by jaslineljy** [reply]
Hi, i am still figuring out how to separate into 4 sources.... however, through the process  i have a few questions that i would like to ask for `Put it all together section`
1. what does the //2+1 mean in `nf = stft_params.window_length // 2 + 1`
2. why is the nac =1 and not 2 since it is stereo audio source. nac was 2 in `Coding up model architectures section` but changed to 1 in `Put it all together section`
3. Under the `Coding up model architectures section` there is a part where inverse stft and putting the phase happens after having the mask. Where does that happen under `Put it all together section` ? Because i don't see it being mentioned but it is an essential step to obtain the correct output isn't it?
4. If we want to use our own random song (`random_audio.wav`) and test the separation of the output, how do we do it? I tried both following ways but both have errors:
```
test_folder = "/content/Test/"
test_data = data.mixer(stft_params, transform=None, 
    fg_path=test_folder, num_mixtures=MAX_MIXTURES, coherent_prob=1.0)
item = test_data[0]

separator.audio_signal = item['mix']
estimates = separator()

viz.show_sources(estimates)
```
```
item = nussl.AudioSignal("/content/Test/TestSeparate.wav")
separator.audio_signal = item
estimates = separator()
viz.show_sources(estimates)
```

But it doesn't make sense to put it here right? since if i want to separate a random track, the random track shouldn't need to have ['mix'].  I am assuming that the nussl.AudioSignal() reads the random track? so how should we exactly assign `random_audio.wav` or separation?
```
separator = nussl.separation.deep.DeepMaskEstimation(
    nussl.AudioSignal(), model_path='checkpoints/best.model.pth',
    device=DEVICE,)
```

5. Under `Coding up model architectures section` there are events to determine coherent, pitch shift and time stretch. However, `Put it all together section`, when we use data on the fly, there is only the coherent event. May i know does that mean the data are already processed with pitch shift and time stretch? Or we have to explicitly do it again like how it is done in `Generaring mixtures with Scaper`
```
labels = ['vocals', 'drums', 'bass', 'other']
for label in labels:
    sc.add_event(label=('const', label),                # set the label value explicitly using a constant
                 source_file=('choose', []),            # choose the source file randomly from all files in the folder
                 source_time=('uniform', 0, 7),         # sample the source (stem) audio starting at a time between 0-7
                 event_time=('const', 0),               # always add the stem at time 0 in the mixture
                 event_duration=('const', sc.duration), # set the stem duration to match the mixture duration
                 snr=('uniform', -5, 5),                # choose an SNR for the stem uniformly between -5 and 5 dB
                 pitch_shift=('uniform', -2, 2),        # apply a random pitch shift between -2 and 2 semitones
                 time_stretch=('uniform', 0.8, 1.2))    # apply a random time stretch between 0.8 (faster) and 1.2 (slower)
```

Very sorry for asking so many questions. Thank you for your patience and time.
